src/OptimalControlUtilities/include/BipedalLocomotionControllers/OptimalControlUtilities/OptimizationProblemElements.h

```diff
@@ -0,0 +1,219 @@
+{
+/**
+ * Constraints class handles the equality and inequality constraint for a QP
+ * optimization problem. i.e. \f$ l \le A x \le u \f$ where
+ * \f$ A \f$ is the constraint matrix
+ * \f$ l \f$ and \f$ u \f$ are the lower and upper bound respectively
+ */
+class Constraints
+{
+    /**
+     * EqualityConstraint contains the equality constraints
+     */
+    typedef struct EqualityConstraint
+    {
+        CostFunctionOrEqualityConstraintElement* element; /**< Pointer to the constraint */
+        iDynTree::IndexRange indexRange; /**< Size and offset of the constraint */
+
+        /**
+         * Constructor
+         */
+        EqualityConstraint(CostFunctionOrEqualityConstraintElement* const element,
+                           const iDynTree::IndexRange& indexRange);
+    } EqualityConstraint;
+
+    typedef struct InequalityConstraint
+    {
+        InequalityConstraintElement* element; /**< Pointer to the constraint */
+        iDynTree::IndexRange indexRange; /**< Size and offset of the constraint */
+
+        /**
+         * Constructor
+         */
+        InequalityConstraint(InequalityConstraintElement* const element,
+                             const iDynTree::IndexRange& indexRange);
+
+    } InequalityConstraint;
+
+    std::vector<EqualityConstraint> m_equalityConstrains; /**< Vector containing all the equality
+                                                             constraints */
+    std::vector<InequalityConstraint> m_inequalityConstrains; /**< Vector containing all the
+                                                                 inequality constraints */
+
+    iDynTree::VectorDynSize m_lowerBound; /**< Lower bound */
+    iDynTree::VectorDynSize m_upperBound; /**< Upper bound */
+    iDynTree::MatrixDynSize m_constraintMatrix; /**<  Constraint matrix */
+
+    int m_numberOfVariables; /**< Number of variables */
+    int m_numberOfConstraints{0}; /**< Number of constraint */
+
+    /**
+     * Get the index of the next constraint
+     * @return the index of the next constraint
+     */
+    int getNextConstraintIndex() const;
+
+public:
+    /**
+     * Constructor
+     * @param handler variable handler object
+     */
+    Constraints(const VariableHandler& handler);
+
+    /**
+     * Add a new constraint. The constraint matrix and the bound vector are resized
+     * @param element pointer to the constraint
+     */
+    void addConstraint(CostFunctionOrEqualityConstraintElement* element);
+
+    /**
+     * Add a new constraint. The constraint matrix and the bound vector are resized
+     * @param element pointer to the constraint
+     */
+    void addConstraint(InequalityConstraintElement* element);
+
+    /**
+     * Get (and compute) the lower and the upper bounds
+     * @return pair containing the bounds. The first element is the lower
+     * bound, the second one is the upper bound
+     */
+    std::pair<iDynTree::VectorDynSize&, iDynTree::VectorDynSize&> getBounds();
```

Comment:
Not blocking merge especially if it is actually code used, but I personally don't like the use of pair for this kind of uses, because then you access the elements as `bounds.first` and `bounds.second`, and that is an harm to readability. 
IMHO a custom template pair-structure with `min` and `max` is more readable , see https://arne-mertz.de/2017/03/smelly-pair-tuple/ .

Reply:
```cpp
Do you mean something like this?
template <typename T, typename U> struct Bounds : public std::pair<T, U>
{
    Bounds(const T& min, const U& max)
        : std::pair<T, U>(min, max)
    {
    }

    T& min()
    {
        return this->first;
    }
    const T& min() const
    {
        return this->first;
    }
    U& max()
    {
        return this->second;
    }
    const U& max() const
    {
        return this->second;
    }
};
```

Reply:
Yes, it would be a good example.